org.eclipse.lsp4e/src/org/eclipse/lsp4e/internal/CharsInputStream.java

```diff
@@ -0,0 +1,224 @@
+/*******************************************************************************
+ * Copyright (c) 2024 Sebastian Thomschke and others.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ * Sebastian Thomschke - initial implementation
+ *******************************************************************************/
+package org.eclipse.lsp4e.internal;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.Charset;
+import java.nio.charset.CharsetEncoder;
+import java.nio.charset.CoderResult;
+import java.nio.charset.StandardCharsets;
+import java.util.Objects;
+import java.util.function.IntSupplier;
+
+import org.eclipse.jdt.annotation.Nullable;
+
+public class CharsInputStream extends InputStream {
+	@FunctionalInterface
+	public interface CharsSupplier {
+		char charAt(int index) throws Exception;
+	}
+
+	private enum EncoderState {
+		ENCODING,
+		FLUSHING,
+		DONE
```

Comment:
```suggestion
		FLUSHED
```

Reply:
DONE in this context means finished, i.e. the encoder will not be used to refill buffers. Flushed could mean the encoder switches to encoding again and flushes again.

Reply:
Thanks. I am having some trouble understanding some parts of the logic, as you see :) 

What about adding a comment explaining what the DONE state means?

Reply:
I added some javadoc. I hope it helps :-)